feat: Add sentry logging to backend

RangerMauve opened the pull request:
Closes #1209 

Also closes https://github.com/digidem/comapeo-core/issues/1033

Tested in a separate nodejs-mobile project

- Adds automatic tracing to HTTP
- Saves all logs to sentry
- Doesn't send PII

Is there an existing consent flag for this tracking on the frontend? Should we have a way to opt in/out?

Commit:
feat: Add sentry logging to backend

RangerMauve commented:
I think I'm missing something regarding how to add dependencies to the backend. I saw that the root package.json seemed to depend on comapeo/core so I tried adding sentry at that level, but that seems to be missing in the backend.

However trying to install the dependency inside src/backend/package.json leads to this error:

```
> npm i --save @sentry/node
npm error code 1
npm error path /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild
npm error command failed
npm error command sh -c node install.js
npm error /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:132
npm error     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
npm error           ^
npm error
npm error Error: Expected "0.19.5" but got "0.25.2"
npm error     at validateBinaryVersion (/home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:132:11)
npm error     at /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:285:5
npm error
npm error Node.js v20.18.3
```

Is there a different way to install dependencies or is node 20 not the correct version to use?

achou11 commented:
> I think I'm missing something regarding how to add dependencies to the backend. I saw that the root package.json seemed to depend on comapeo/core so I tried adding sentry at that level, but that seems to be missing in the backend.
> 
> However trying to install the dependency inside src/backend/package.json leads to this error:
> 
> ```
> > npm i --save @sentry/node
> npm error code 1
> npm error path /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild
> npm error command failed
> npm error command sh -c node install.js
> npm error /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:132
> npm error     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got ${JSON.stringify(stdout)}`);
> npm error           ^
> npm error
> npm error Error: Expected "0.19.5" but got "0.25.2"
> npm error     at validateBinaryVersion (/home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:132:11)
> npm error     at /home/mauve/src/comapeo-mobile/src/backend/node_modules/esbuild/install.js:285:5
> npm error
> npm error Node.js v20.18.3
> ```
> 
> Is there a different way to install dependencies or is node 20 not the correct version to use?

We have a [test](https://github.com/digidem/comapeo-mobile/blob/a0806aa6609de7798521f309657dcc37cafb19c6/src/frontend/packageJson.test.ts) that does some checks on the frontend and backend package.json files (due to some strict requirements we want to enforce).

The error you're seeing is because you're using a non-exact version of `@sentry/node` at the _top-level_ package.json. From what I can tell, you're only intending to install it in the backend, in which case you only need to install the module via the `src/backend/package.json` and leave the top-level one alone. We don't enforce a strict version for deps used in the backend from what I remember. The test only makes sure that the relevant core-related deps between top-level and backend match

achou11 commented:
sorry, i was commenting on the error that CI was raising. Not super sure about the esbuild-related error when trying to install in the backend though 🤔 I was able to install it locally without any issue (I'm using node 20.19.2).

achou11 commented:
> I saw that the root package.json seemed to depend on comapeo/core

Core is installed as a _dev_ dependency at the root level with the intent of guaranteeing the same version that's used as prod dep in the backend in order for the types to align when using other core-related deps in the frontend code e.g. schema, ipc, and core-react.

in general, my loose guide for installing new deps:

- If it's purely used in the frontend e.g. react native application or tooling related to that, update the root package.json

- If it's purely used in the backend, update the backend package.json. if the dep is core-related, there's a good chance that the root level package.json also needs to be updated in order for correct resolution of types

- If it's used in both, update both package.json files.

Commit:
chore: Move sentry node to backend package

gmaclennan reviewed (changes requested):
Looks good so far, and I think there are a few other integrations to make with the frontend. Initially I think we can do all of these as startup flags for the node process by passing them to [`initializeNodejs()`](https://github.com/digidem/comapeo-mobile/blob/e199928271b45cd6b874607e217d768617aeaf79/src/frontend/initializeNodejs.ts#L8-L9) and setting them as startup flags, which are then [parsing in the backend](https://github.com/digidem/comapeo-mobile/blob/e199928271b45cd6b874607e217d768617aeaf79/src/backend/index.js#L19-L20):

- [`sentryEnvironment`](https://github.com/digidem/comapeo-mobile/blob/e199928271b45cd6b874607e217d768617aeaf79/src/frontend/App.tsx#L41-L47), which helps us separate noise from dev and QA from production events.
- The same user ID for [`initialScope`](https://docs.sentry.io/platforms/javascript/guides/node/configuration/options/#initialScope) as defined by [`getSentryUserID()`](https://github.com/digidem/comapeo-mobile/blob/e199928271b45cd6b874607e217d768617aeaf79/src/frontend/App.tsx#L55-L56) - that will help us match frontend and backend events. It's a random rotating ID - by default Sentry creates a persistent random ID (in the frontend at least)
- [`metricsIsEnabled`](https://github.com/digidem/comapeo-mobile/blob/e199928271b45cd6b874607e217d768617aeaf79/src/frontend/App.tsx#L123) - note that this will mean changes come into effect only when the app restarts, but I think that's an ok compromise for now. We should set the init option [`enabled`](https://docs.sentry.io/platforms/javascript/guides/node/configuration/options/#enabled) based on this.

RangerMauve commented:
I'm still getting a weird error where it says it can't find `@sentry/node` depsite it being in the `src/backend/package.json` and the `src/backend/node_modules`. I ran the install with `--ignore-scripts`. Do I need to add the module in the build somewhere or maybe use a different method to import it?

achou11 commented:
> I'm still getting a weird error where it says it can't find `@sentry/node` depsite it being in the `src/backend/package.json` and the `src/backend/node_modules`. I ran the install with `--ignore-scripts`. Do I need to add the module in the build somewhere or maybe use a different method to import it?

TL;DR - you should either:

1. Move the Sentry code you've added to `src/backend/index.js` (not `loader.js`)

2. Update the [script](https://github.com/digidem/comapeo-mobile/blob/feat/initial-backend-logs/scripts/build-backend.mjs) that builds the backend to make sure that `@sentry/node` isn't purged from the included `node_modules` directory that's loaded into the android assets.

My guess is that (1) will be preferable but I don't have full context.

---

When building the backend for the app we [bundle the code using the `src/backend/index.js` file as the entry point](https://github.com/digidem/comapeo-mobile/blob/8bc4a1e80646d8886ac19b07413c08e07437658b/src/backend/package.json#L9). This is for performance + space reasons, as we want to avoid having to include a full `node_modules` directory loaded into the android app. The `loader.js` is not transformed and ran as-is by nodejs-mobile when starting the backend.

We still have a `node_modules` directory that's included as part of the assets loaded into the app, but we purge it and only keep things that are needed for static resolution e.g. maps or categories files ([link](https://github.com/digidem/comapeo-mobile/blob/8bc4a1e80646d8886ac19b07413c08e07437658b/scripts/build-backend.mjs#L96-L116))

RangerMauve commented:
I'll try (1) but I worry it won't work as well because sentry needs us to run the init before everything else so it can attach itself onto dependencies

achou11 commented:
> I'll try (1) but I worry it won't work as well because sentry needs us to run the init before everything else so it can attach itself onto dependencies

hm yeah I see the note in their setup guide about that. part of me wonders if we can get rid of the separate `loader.js` and just move that initialization code into the `index.js`.

EDIT: we tell nodejs-mobile which script to run [here](https://github.com/digidem/comapeo-mobile/blob/8bc4a1e80646d8886ac19b07413c08e07437658b/src/frontend/initializeNodejs.ts#L24) - could just be a matter of pointing it to the bundled index.js file

gmaclennan commented:
> hm yeah I see the note in their setup guide about that. part of me wonders if we can get rid of the separate `loader.js` and just move that initialization code into the `index.js`.

The separate loader code was copied over from the original manyverse code in Mapeo Legacy. I believe it was needed so that `os.homedir()` and `process.cwd()` are defined before other modules load. I'm not sure if it's needed for our code as it is right now.

gmaclennan committed:
Update bundle script for Sentry

This creates a separate bundle for loader.js
(which includes the Sentry SDK) and loads index.js
dynamically so that Sentry can instrument it.

This also creates a bundle for the module hooks
that Sentry uses to instrument modules, because
rollup does not know how to bundle module hooks.

Commit:
fix: Enable mapeo debug logs in backend

gmaclennan committed:
Fix bundling issues with new rollup config

Switching to a rollup config file, rather than
using the rollup javascript API directly, allows
rollup to automate a few things e.g. creating
shared chunks and ensuring the source maps are
correct.

gmaclennan committed:
Small cleanup and adding docs/comments

Commit:
fix: capture debug logs with sentry via console log

Commit:
feat: pass sentry config from frontend to backend

RangerMauve commented:
TODOs:

- only log errors on prod
- double check redactions for ids.

Commit:
feat: add user id to console log sentry traces

RangerMauve commented:
We don't have a way to distinguish log level in the core logger ATM so I'll make logging a flag.

https://github.com/digidem/comapeo-core/blob/main/src/logger.js

Commit:
feat: ignore console logs in prod

Commit:
chore: fix up ts errors for backend opts

Commit:
fix: account for parseArgs improperly handling booleans

gmaclennan commented on src/backend/loader.js:
```diff
@@ -14,4 +16,54 @@ os.homedir = () => nodejsProjectDir
+    sentryUserId: { type: 'string' },
+    metricsIsEnabled: { type: 'string' },
+  },
+  strict: false,
```
This is ok as-is, but just to note that we could change remove `strict: false` and avoid the typeof checks below - we would need to add rootKey parsing here (must parse all arguments) and we would need to change the arg for metricsIsEnabled to an actual boolean arg (e.g. true if present, rather than a string arg with value `true|false`).

RangerMauve replied:
I think I'd prefer to keep it as is so we don't need to keep the two sets of arguments in sync. Another option could be to extract the argument parsing into a separate module to import into both places. Is it safe to import libraries in the loader like that or would it mess with the bundler?

gmaclennan replied:
Yes the bundler should extract it as a separate file that is imported by both

gmaclennan commented on src/frontend/App.tsx:
```diff
@@ -133,6 +133,9 @@ persistedMetricsDiagnosticsStore.instance.subscribe((current, previous) => {
 });
 
+// Need to know if metrics are enabled before starting node
+initializeNodejs(metricsIsEnabled, sentryEnvironment, sentryUserId);
```
nitpick, but in general our pattern has been to pass an object, rather than rely on positional parameters. It's slightly mute with Typescript, but in general passing `options` as an object removes the chance of bugs from getting the parameter positions mixed up (this has happened to me plenty)

gmaclennan commented on src/backend/loader.js:
```diff
@@ -14,4 +16,54 @@ os.homedir = () => nodejsProjectDir
+const sentryDebug = sentryEnvironment === 'development'
+const initialScope = sentryUserId ? { user: { id: sentryUserId } } : undefined
+
+const logLevels = ['error']
```
This removes the need for @ts-expect-error below - it ensures that these levels are typed as string literals, rather than just a generic string.

```suggestion
/** @type {Array<"error" | "log" | "warn">} */
const logLevels = ['error']
```

gmaclennan commented on src/backend/loader.js:
```diff
@@ -14,4 +16,54 @@ os.homedir = () => nodejsProjectDir
+    },
+  },
+  tracesSampleRate: 1.0,
+  // @ts-expect-error Unclear where to import type from their docs
```
```suggestion
```

gmaclennan commented on src/backend/src/app.js:
```diff
@@ -24,6 +25,13 @@ const MAPBOX_ACCESS_TOKEN =
+// needed to have logs captured by sentry
+debug.log = function log(...args) {
+  // @ts-expect-error InspectOpts shouldn't be undefined but might be
+  return console.log(formatWithOptions(debug.inspectOpts, ...args))
```
This change fixes the undefined issue. We still need the @ts-expect-error, because the way [debug parses env variables](https://github.com/debug-js/debug/blob/33330fa8616b9b33f29f7674747be77266878ba6/src/node.js#L124-L145) means that the opts could have incorrect values, but I think it's ok to skip validating that since we are not setting them from env at this time.

```suggestion
  return console.log(formatWithOptions(debug.inspectOpts || {}, ...args))
```

gmaclennan commented on src/backend/loader.js:
```diff
@@ -14,4 +16,54 @@ os.homedir = () => nodejsProjectDir
+  environment: sentryEnvironment,
+  initialScope,
+  _experiments: {
+    enableLogs: true,
```
I think we should only enable logs if `sentryEnvironment !== 'production'`, e.g. only for dev and QA for now - until we have added the consent opt-in for "help make comapeo better".

gmaclennan reviewed (changes requested):
Some very minor suggestions, but the one thing to change before we merge is to only capture logs in QA and dev for now.

RangerMauve committed:
Update src/backend/loader.js

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

RangerMauve committed:
Update src/backend/loader.js

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

RangerMauve committed:
Update src/backend/src/app.js

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

Commit:
fix: only enable logs when not production

Commit:
chore: Use object for params to initializeNodejs

Commit:
core: unify backend arg parsing

Commit:
core: unify backend arg parsing

gmaclennan commented on src/backend/src/args.js:
```diff
@@ -0,0 +1,35 @@
+    throw new Error('backend did not receive sentryEnvironment')
+  if (typeof values.rootKey !== 'string')
+    throw new Error('backend did not receive root key from front end')
+  if (typeof values.version === 'string')
```
```suggestion
  if (typeof values.version !== 'string')
```

gmaclennan replied:
This will fix the typecheck that is failing in CI.
```suggestion
  if (typeof version === 'string')
```

gmaclennan commented on src/backend/src/args.js:
```diff
@@ -0,0 +1,35 @@
+    throw new Error('backend did not receive sentryUserId')
+  if (typeof sentryEnvironment !== 'string')
+    throw new Error('backend did not receive sentryEnvironment')
+  if (typeof values.rootKey !== 'string')
```
```suggestion
  if (typeof rootKey !== 'string')
```
This will fix the typecheck failing CI - it ensures that the returned value has a type-guard to prevent it being `undefined`.

gmaclennan reviewed (changes requested):
Some small changes needed, go ahead and merge without further review after fixing.

Commit:
fix: Properly typecheck backend args

Commit:
fix: Proper check for boolean metricsIsEnabled

Commit:
fix: Don't check for version string

RangerMauve commented:
All changes addressed, CI is passing. Merging
